src/java.base/share/classes/java/security/MessageDigest.java

```diff
@@ -307,6 +308,147 @@ public static MessageDigest getInstance(String algorithm,
+     *         specified algorithm
+     *
+     * @throws IllegalArgumentException if the provider name is {@code null}
+     *         or empty
+     *
+     * @throws NoSuchAlgorithmException if a {@code MessageDigestSpi}
+     *         implementation for the specified algorithm and parameters
+     *         is not available from the specified provider
+     *
+     * @throws NoSuchProviderException if the specified provider is not
+     *         registered in the security provider list
+     *
+     * @throws NullPointerException if {@code algorithm} is {@code null}
+     *
+     * @see Provider
+     * @since 22
+     */
+    public static MessageDigest getInstance(String algorithm,
+            AlgorithmParameterSpec params, String provider)
+            throws NoSuchAlgorithmException,
+                    NoSuchProviderException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        GetInstance.Instance instance = GetInstance.getInstance("MessageDigest",
+                MessageDigestSpi.class, algorithm, params, provider);
+        return Delegate.of((MessageDigestSpi)instance.impl,
+                algorithm, instance.provider);
+    }
+
+    /**
+     * Returns a {@code MessageDigest} object that implements the specified
+     * digest algorithm and is initialized with the specified parameters.
+     *
+     * <p> This method traverses the list of registered security Providers,
+     * starting with the most preferred Provider.
+     * A new {@code MessageDigest} object encapsulating the
+     * {@code MessageDigestSpi} implementation from the first
+     * provider that supports the specified algorithm is returned.
+     *
+     * <p> Note that the list of registered providers may be retrieved via
+     * the {@link Security#getProviders() Security.getProviders()} method.
+     *
+     * @implNote
+     * The JDK Reference Implementation additionally uses the
+     * {@code jdk.security.provider.preferred}
+     * {@link Security#getProperty(String) Security} property to determine
+     * the preferred provider order for the specified algorithm. This
+     * may be different from the order of providers returned by
+     * {@link Security#getProviders() Security.getProviders()}.
+     *
+     * @param algorithm the name of the algorithm requested.
+     * See the MessageDigest section in the <a href=
+     * "{@docRoot}/../specs/security/standard-names.html#messagedigest-algorithms">
+     * Java Security Standard Algorithm Names Specification</a>
+     * for information about standard algorithm names.
+     * @param params the initialization parameters (may be {@code null}).
+     *
+     * @return a {@code MessageDigest} object that implements the
+     *         specified algorithm
+     *
+     * @throws NoSuchAlgorithmException if no {@code Provider} supports a
+     *         {@code MessageDigestSpi} implementation for the
+     *         specified algorithm and parameters
+     *
```

Comment:
Hmm, interesting. I have a slight preference for the `CertStore` approach because it feels like the caller can more easily distinguish between the two cases but I could probably live with either. What do you prefer?